### ISSUE TITLE
GIT_EXEC_PATH is set in terminal's environment

### ISSUE THREAD

**Issue by perrinjerome** [eclipse-theia/theia#7555]
### Description

In the terminal from theia, `GIT_EXEC_PATH` environment variable seem to be set implicitly.

This can cause issues when trying to use another git version in this terminal.

I guess this is set by this code:

https://github.com/eclipse-theia/theia/blob/22eb3a28b78233963ac3f5a344336ac8829c7369/packages/git/src/node/init/git-init.ts#L75

### Reproduction Steps

 * Start theia with git `2.23.0` in `$PATH`
 * Install git `2.26.0` in another location (using `--prefix` to another directory)
 * set `$PATH` to include bin directory of the prefix where git `2.26.0` is installed.
 * use this git command:
```
$ git --version
git version 2.26.0
$  git  clone --recursive -v --progress https://github.com/eclipse-theia/theia/
Cloning into 'theia'...
POST git-upload-pack (gzip 10065 to 5001 bytes)
remote: Enumerating objects: 2881, done.
remote: Counting objects: 100% (2881/2881), done.
remote: Compressing objects: 100% (182/182), done.
remote: Total 70165 (delta 2742), reused 2789 (delta 2694), pack-reused 67284
Receiving objects: 100% (70165/70165), 126.14 MiB | 22.96 MiB/s, done.
Resolving deltas: 100% (50539/50539), done.
usage: git submodule [--quiet] [--cached]
   or: git submodule [--quiet] add [-b <branch>] [-f|--force] [--name <name>] [--reference <repository>] [--] <repository> [<path>]
   or: git submodule [--quiet] status [--cached] [--recursive] [--] [<path>...]
   or: git submodule [--quiet] init [--] [<path>...]
   or: git submodule [--quiet] deinit [-f|--force] (--all| [--] <path>...)
   or: git submodule [--quiet] update [--init] [--remote] [-N|--no-fetch] [-f|--force] [--checkout|--merge|--rebase] [--[no-]recommend-shallow] [--reference <repository>] [--recursive] [--] [<path>...]
   or: git submodule [--quiet] set-branch (--default|--branch <branch>) [--] <path>
   or: git submodule [--quiet] summary [--cached|--files] [--summary-limit <n>] [commit] [--] [<path>...]
   or: git submodule [--quiet] foreach [--recursive] <command>
   or: git submodule [--quiet] sync [--recursive] [--] [<path>...]
   or: git submodule [--quiet] absorbgitdirs [--] [<path>...]
```
This error is because the `git` command from `$PATH` (`2.26.0`) is using `git-submodule` from `$GIT_EXEC_PATH` ( `2.23.0`) and they are incompatible.

Once I figured this out, the workaround is easy, when I adjust my `$PATH` to use new git version, I also adjust `$GIT_EXEC_PATH` accordingly, but it took me some some time to understand what was going on and I was thinking that it would be better if this environment variable was not set in terminal processes.

**OS and Theia version:**

theia 0.12.0-next.f255f5a8 on linux 

**Diagnostics:**

**Comment by akosyakov** [comment]
@AlexTugarev @kittaakos maybe you can help here?

**Comment by kittaakos** [comment]
@perrinjerome, can you tell us what kind of issue are you trying to solve? Why do not you install Git `2.26.0` before starting the Theia backend?

I can see, you have already checked the implementation, did you see this message too? Can it be that this has caused an issue on your side?

https://github.com/eclipse-theia/theia/blob/22eb3a28b78233963ac3f5a344336ac8829c7369/packages/git/src/node/init/git-init.ts#L59-L60

**Comment by perrinjerome** [comment]
My use case is very specific, I'm using theia as a development environment for a script which downloads a more recent version of git, compile and installs it in a "user folder" ( using `--prefix` to configure ) and then use this new git to clone a repository.

When I ran this script in theia's terminal I had this problem because this `GIT_EXEC_PATH` environment variable is set in the terminal.

If we say it's a bug in my script and I should unset `GIT_EXEC_PATH` in my script, that's fine for me, but I was thinking that it would generally be better if theia set `GIT_EXEC_PATH` not as an environment variable that is set to current process, because all terminals, all tasks etc will have it and there might be some cases like mine where it cause problems, even if the probability is low.

I have not looked at the code much, but I see that there's a "git environment" around here:
https://github.com/eclipse-theia/theia/blob/ba717c52e7d00d102ee170e6e5b62bc796366082/packages/git/src/node/env/git-env-provider.ts#L29
Isn't this to set environment variables to the `git` process ? If it's easy to use this to set `GIT_EXEC_PATH` only to child git processes and not to all processes, that would be better I suppose.

**Comment by kittaakos** [comment]
Thanks for writing down the details, @perrinjerome!

A few things:
 - `GIT_EXEC_PATH` (and `LOCAL_GIT_DIRECTORY` too) are internal things and used by a downstream dependency that we use to run Git commands. These commands are associated with the UI and not the terminal.
 - When you have terminal and you execute and Git commands, it does nothing with the `GIT_EXEC_PATH` env variable, but it uses the Git which is available on system-level.
 - When you start a new terminal and execute `set` you can see all env variables. Among them, you can see `GIT_EXEC_PATH` too, it is a side-effect of changing info between processes via an env var.

I am still trying to understand what you are trying to achieve, so let me recap and you can correct me.
 - You have version `X` Git on the system by default.
 - I assume when you start Theia and open a terminal, and you execute `git version` you can see version `X` of Git. You can clone too. So far good.
 - You want to install version `Y` Git (where version `Y` > version `X`).
 - You install Git into your home folder or wherever it fits your needs.
 - You open a terminal and try to execute the command `git  clone --recursive -v --progress https://github.com/eclipse-theia/theia/` with version `Y` and it fails. Is this what has happened?

What does it print when you execute `which git` in one of Theia's terminals?

**Comment by perrinjerome** [comment]
Thanks for feedback. I tried to make a minimal reproduction running on gitpod:

https://gitpod.io/#https://github.com/perrinjerome/theia_repro_7555

this image uses git 2.26 and inside the workspace, there's a script which installs another git version, 2.24 and then uses it. Note that it's different from my original reproduction steps, I started with 2.24 and then installed 2.26 in the workspace, but it's still a scenario which leads to using a git command with an incompatible `GIT_EXEC_PATH`. This time the error message is different, *warning: You appear to have cloned an empty repository.*

![image](https://user-images.githubusercontent.com/521510/79300116-c3ddde80-7f20-11ea-8df2-8cd402156336.png)

It's not a common use case I guess, but I this shows that it would be better if environment variables that are needed for the git extension (  `GIT_EXEC_PATH` ) were only set in sub-processes spawned by the git extension and not in the "main" process, because terminal processes also inherit the environment variables and it can cause surprising behaviours.

I'm not sure I understand the code well, but I guess instead of setting the environment variables in the current process here:

https://github.com/eclipse-theia/theia/blob/22eb3a28b78233963ac3f5a344336ac8829c7369/packages/git/src/node/init/git-init.ts#L74-L75

we change to set this environment in the `GitEnvProvider` here:

https://github.com/eclipse-theia/theia/blob/d40e3ddf0daf912a7ce00034d72d149ac6a22f7d/packages/git/src/node/env/git-env-provider.ts#L47

and probably same for electron version here:

https://github.com/eclipse-theia/theia/blob/d40e3ddf0daf912a7ce00034d72d149ac6a22f7d/packages/git/src/electron-node/env/electron-git-env-provider.ts#L40-L45

then `GIT_EXEC_PATH` would only be set for git commands spawned by the git extension and there will be no magic happening in other processes.

**Comment by kittaakos** [comment]
Ohh, I did not know you have issues with Git in Gitpod. @AlexTugarev, can you please help with this? Thank you!

**Comment by akosyakov** [comment]
@perrinjerome Could you open the issue in https://github.com/gitpod-io/gitpod instead? I am not sure that anything can be done in Theia about it. It seems to be related to how Theia is integrated in Gitpod.

**Comment by perrinjerome** [comment]
Thank you, I don't have this issue as a user of gitpod. I noticed this with a theia installed locally, it also happens on theia-apps images and on gitpod, I used gitpod here to show a minimal reproduction.

**Comment by AlexTugarev** [comment]
There isn't a Gitpod specific problem. As @perrinjerome pointed out, `GitInit` is patching the `process.env` to enable git integration in Theia, but then this is inherited to terminal processes as well. Actually, this ensures you are using the same tool.

Maybe we should unset `GIT_EXEC_PATH` for terminal processes. WDYT?

**Comment by kittaakos** [comment]
> Maybe we should unset `GIT_EXEC_PATH` for terminal processes. WDYT?

Yes, we can. But I don't understand how `GIT_EXEC_PATH` would affect Git from the `$PATH`.

@perrinjerome, I have checked your `.sh` script once more:
 - Are you positive you build Git from the sources correctly,
 - `GIT_EXEC_PATH` is not the same as the Git from the `$PATH`. If you want to get `GIT_EXEC_PATH` you have to run Git from the `$PATH` with the `--exec-path` flag.

**Comment by perrinjerome** [comment]
I looked a bit more at all this, and it's really dugite which expect `GIT_EXEC_PATH` environment variable to be set for current process and there does not seem to be an easy way to tell it to only set `GIT_EXEC_PATH` to child git process it may spawn, so that idea with `GitEnvProvider` seems not really possible.

>  But I don't understand how GIT_EXEC_PATH would affect Git from the $PATH.

I was not clear in my explanations on this part I think. As you know, `git` command uses "helpers", so when you type for example `git submodule args` git will lookup for a binary called `git-submodule` in git's exec path.

The  `git  clone --recursive` uses `git submodule` internally and it seems to pick a `git-submodule` not from the git command installed locally but from the `GIT_EXEC_PATH` from environ and both versions seem to be incompatible.

In a sense it's my fault, because if I want to use a git installation from a different `$PATH` I also have to adjust the `$GIT_EXEC_PATH` to also use the same git. But what was confusing it's that I never set `$GIT_EXEC_PATH` myself. So maybe  that's  "not a bug" and in any case, I don't see this a high priority thing.

>> Maybe we should unset `GIT_EXEC_PATH` for terminal processes. WDYT?
> For completeness; we should remove all env variables from the terminal, shouldn't we?

Yes, that could be a way, also maybe do the same for tasks processes.

Not all variables can be unset though,  for example if `$PATH` is unset it would not be possible to execute commands. Also probably some extensions are relying on the current behaviour that they can modify `process.env` and that it's reflected in terminal processes. For example, I see gitpod sets a `GIT_EDITOR` environment variable so that git in the terminal use an editor pane from theia. This is probably achieved by changing `process.env`( and it's a super cool feature by the way ;) )

**Comment by akosyakov** [comment]
> For completeness; we should remove all env variables from the terminal, shouldn't we?
> Yes, that could be a way, also maybe do the same for tasks processes.

I don't think it is a way. End users configured own shell and start Theia from it. It is expected that terminals started by Theia using the same shell configuration to have tools. Removing env variables will break everybody. I wonder could not end user configure env variables in the shell that `$PATH` and `$GIT_EXEC_PATH` are consistent?

**Comment by sgraband** [comment]
Caused by `@theia/git`'s `git-init.ts` which is deprecated. The VS Code built-in Git extension does not set this variable.
